---
title: "Functional analysis of multi-omics data"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2); theme_set(cowplot::theme_cowplot(font_size = 15))
library(reshape2)
library(ggrepel)
library(knitr)
library(Biostrings)
library(RColorBrewer)
library(here)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
```

```{r}
options(ggplot2.discrete.colour= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
```

# 1. Load data
```{r}
diff_results <- read_tsv("data/processed_data/diff_results_20230322.tsv", show_col_types = FALSE)
```


# 2. Overlapping hits 

## filter secreted factors
```{r}
# secreted proteins only 
secreted_proteins <- msigdbr::msigdbr(species = "Homo sapiens") %>% 
  filter(gs_id %in% c('M5889', 'M5885')) #NABA_MATRISOME, #NABA_MATRISOME_ASSOCIATED 
```

## heatmap overlapping hits 
```{r}

# set cut offs
lfc_cutoff <- log2(1.5)
p_cutoff <- 0.05

overlap <- diff_results %>%
  subset(abs(logFC) >= lfc_cutoff & adj.P.Val <= p_cutoff) %>%
  filter(!(modality == "rna" & abs(logFC) < log2(2))) %>% # logfc cut off for rna 
  filter(!(modality == "secretomics" & !(feature_id %in% secreted_proteins$gene_symbol))) %>% 
  filter(modality != "phospho" & modality != "input_phospho") %>%
  # mutate(modality = ifelse(modality == "rna", "rna", "protoemics technologies")) %>%
  # modify feature_id to be able to collapse phospho
  ungroup() %>%
  distinct(modality, feature_id)%>% 
  group_by(feature_id) %>% 
  summarise(n_hits = n_distinct(modality)) %>% 
  filter(n_hits >1) %>% 
  bind_rows(data.frame(feature_id = c("MARCKS", "CDH11", "NRP2", "SDC4", "TFPI", "CPA4", "CACHD1"))) # add secreted factors that were previously excluded

# define order for columns
col_order <- c(
  "12h_secretomics", "24h_secretomics", "48h_secretomics", "72h_secretomics", "96h_secretomics",
  "12h_proteomics", "24h_proteomics", "48h_proteomics", "72h_proteomics", "96h_proteomics",
  "1h_rna", "12h_rna", "24h_rna", "48h_rna", "72h_rna", "96h_rna"
  )

set.seed(2)

p2 <- diff_results %>% 
  filter(modality != "phospho" & modality != "input_phospho") %>%  #input
  filter(feature_id %in% overlap$feature_id) %>% #x
  mutate(time = factor(time, levels = c("0h", "0.08h", "1h", "12h","24h", "48h", "72h", "96h")),
          hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  filter(hit == "hit") %>% 
  acast(feature_id ~ time + modality, value.var = "logFC", fill = 0) %>% 
  ComplexHeatmap::Heatmap(
  cluster_columns = F, 
  show_row_dend = F,
  show_column_dend = F,
  show_row_names = T,
  row_km = 3,
   row_names_gp = grid::gpar(fontsize = 8.5),  
  column_names_gp = grid::gpar(fontsize = 11),
    rect_gp = grid::gpar(col = "lightgrey", lwd = 0.1),
  column_order = col_order,
  col = circlize::colorRamp2(c(-4, 0, 4), c("dodgerblue4", "white","darkred")),
  name = "log2FC"
  )

p2


```



# 3. Top hits phosho

```{r}
hits <- diff_results %>%filter(modality == "input_phospho") %>%  mutate( hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%   filter(hit == "hit")%>%
  arrange(-abs(logFC)) %>%
  head(57)

data <- diff_results %>%
  mutate( hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  filter(modality == "input_phospho" & hit == "hit") %>%
  filter(feature_id %in% hits$feature_id) %>%
  mutate(time = factor(time, levels = c("0h", "0.08h", "1h", "12h","24h", "48h", "72h", "96h")),
          hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  acast(feature_id ~ time + modality, value.var = "logFC", fill = 0)

set.seed(14)

data %>%
  ComplexHeatmap::Heatmap(
    na_col = "white",
    # km = 3,
    cluster_columns = F,
    show_row_dend = F,
    show_column_dend = F,
    show_row_names = T,
    row_km = 4,
    row_names_gp = grid::gpar(fontsize = 8.5),  # Adjust font size here
    column_names_gp = grid::gpar(fontsize = 11),
    rect_gp = grid::gpar(col = "lightgrey", lwd = 0.1), ##3A3A3A
    col = circlize::colorRamp2(c(-2, 0, 2), c("dodgerblue4", "white", "darkred")),
    name = "log2FC"
  )


```


# 4. heatmap for genes of interest 
as an example I am plotting the PC loading vectors of the transcriptomics data
```{r}
genes <- c("SLC7A14", "FAM107A", "RGCC", "GPNMB", "ABCB5", "PLA1A", "KCP", "HLF", "CHI3L1", "ACP5", "EVI2A", "IGFN1", "HRK", "USH2A", "GPNMB", "IFITM10", "GPRIN3", "COL1A2", "IGFL2-AS1", "LINC01711", "COL6A3", "RASGRF2", "MSC", "FAP", "NKAIN4", "LDLRAD4", "SFRP2", "FOXS1", "FLJ16779", "POSTN", "ACTC1", "ID3", "ID1", "SAMD11", "LY96", "PSG9", "RENBP", "FOLR3")

overlap <- diff_results %>%
  filter(modality == "rna") %>% 
  filter(feature_id %in% genes)%>%  
  mutate( hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%   filter(hit == "hit")

p_pc <- overlap %>% 
  filter(modality == "rna") %>%  #input
  filter(feature_id %in% overlap$feature_id) %>% #x
  mutate(time = factor(time, levels = c("0h", "0.08h", "1h", "12h","24h", "48h", "72h", "96h")),
          hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  filter(hit == "hit") %>% 
  acast(feature_id ~ time + modality, value.var = "logFC", fill = 0) %>% 
  ComplexHeatmap::Heatmap(
      na_col = "white",
    # km = 3,
    cluster_columns = F,
    show_row_dend = F,
    show_column_dend = F,
    show_row_names = T,
    row_names_gp = grid::gpar(fontsize = 8.5),  # Adjust font size here
    column_names_gp = grid::gpar(fontsize = 11),
    rect_gp = grid::gpar(col = "lightgrey", lwd = 0.1), ##3A3A3A
    col = circlize::colorRamp2(c(-4, 0, 4), c("dodgerblue4", "white", "darkred")),
    name = "log2FC"
  )

p_pc

```


### TFs and kinases

```{r}
# List of kinases
kinases <- c("AURKB", "CAMK2D", "CDK3", "CHEK1", "CSNK2A1", "CSNK2A2", "PLK3", "PPP2CB", "PRKCZ", "RAF1", "ROCK1", "ROCK2", "UHMK1", "AKT1", "AKT3", "MAP2K1", "MAPKAPK2", "PRKD1", "PRKG1", "PRKG2", "RPS6KA1", "RPS6KB1", "MET", "PTK2", "PTPRG", "ABL1", "MAP3K8", "MAPK1", "RET", "PDK1", "PRKCG", "SRC", "CDK6", "PPP2CA")

# Row order for kinases
kin_order <- c("PTK2", "RET", "MAP3K8", "MET", "MAPK1", "ABL1", "SRC", "CSNK2A2", "CSNK2A1", "MAPKAPK2", "MAP2K1", "PRKG2", "PRKD1", "PRKG1", "RPS6KB1", "AKT3", "PRKCG", "PRKCZ", "ROCK1", "PTPRG", "ROCK2", "CDK6", "PPP2CA", "CHEK1", "PLK3", "PPP2CB", "CDK3", "UHMK1", "AKT1", "RPS6KA1", "CAMK2D", "AURKB", "PDK1", "RAF1")

# overlap <- diff_results %>%
#   filter(modality %in% c("rna", "proteomics")) %>% 
#   filter(feature_id %in% c(kinases))#%>%  
#  # mutate( hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%   filter(hit == "hit")

column_order <- c("0.08h_proteomics", "1h_proteomics", "12h_proteomics", "24h_proteomics","48h_proteomics",
                  "72h_proteomics", "96h_proteomics", "0.08h_rna", "1h_rna", "12h_rna", "24h_rna", "48h_rna", 
                  "72h_rna", "96h_rna")


set.seed(6)
p_kin <- diff_results %>% 
  filter(modality %in% c("rna","proteomics")) %>%  #input
  filter(feature_id %in% kinases) %>% #x
  #mutate(time = factor(time, levels = c("0h", "0.08h", "1h", "12h","24h", "48h", "72h", "96h")))%>%#,
          #hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  #filter(hit == "hit") %>% 
  acast(feature_id ~ time + modality, value.var = "logFC", fill = 0) %>% 
  ComplexHeatmap::Heatmap(
      na_col = "white",
    # km = 3,
    cluster_columns = F,
    show_row_dend = F,
    show_column_dend = F,
    show_row_names = T,
    #cluster_rows = F, # enable for alphabetical ordering
    row_names_gp = grid::gpar(fontsize = 8.5),  # Adjust font size here
    column_names_gp = grid::gpar(fontsize = 11),
    rect_gp = grid::gpar(col = "lightgrey", lwd = 0.1), ##3A3A3A
    col = circlize::colorRamp2(c(-2, 0, 2), c("dodgerblue4", "white", "darkred")),
    name = "log2FC",
    column_order = column_order
  )

p_kin



Tfs <- c("SMAD4", "ETS2", "SRF", "KLF1", "EPAS1", "ATF6", "FOXO4", "GLI1", "KMT2A", "HIC1", "TEAD4", "FOSL2", "ONECUT1", "ZBTB7A", "SMAD5", "KLF9", "E2F1", "RARA", "EGR1", "ATF3", "NR5A1", "FLI1", "HIF1A", "TFAP2A", "SMAD1", "NR4A1", "NR2F2", "FOXL2", "BHLHE40", "FOXK2", "MNT", "THAP1", "ZKSCAN1", "TP53", "ETS1", "PAX6", "NFKB1", "ATF4", "RFX1", "SREBF1", "STAT2", "RFX5", "BHLHE22", "TP63", "PAX8", "MITF", "ELF3", "FOXA1", "HNF4G", "LHX2", "SOX11", "STAT4", "MYC", "E2F4", "FOXM1")

TF_order <- c("SMAD5", "SMAD4","EPAS1","SRF", "TP53","KLF9", "ATF6","TEAD4","ETS1", "PAX6","BHLHE22","TP63", "SOX11","KMT2A", "FOSL2","ZKSCAN1", "RFX1", "THAP1","FOXL2","NFKB1", "EGR1","BHLHE40","NR4A1","MNT", "E2F1", "SMAD1","FLI1","RARA","NR2F2","HIF1A","FOXK2", "STAT4","ATF3","ATF4","HNF4G","FOXA1", "FOXM1", "E2F4", "STAT2", "GLI1","SREBF1","NR5A1","TFAP2A",  "ETS2","HIC1", "FOXO4", "MYC","ZBTB7A","MITF","KLF1","ELF3","LHX2","ONECUT1", "RFX5", "PAX8")


set.seed(6)
p_tf <- diff_results %>% 
  filter(modality %in% c("rna","proteomics")) %>%  #input
  filter(feature_id %in% Tfs) %>%#,
          #hit = ifelse(adj.P.Val < 0.05 & abs(logFC) > log2(1.5), "hit", "no hit")) %>%
  #filter(hit == "hit") %>% 
  acast(feature_id ~ time + modality, value.var = "logFC", fill = 0) %>% 
  ComplexHeatmap::Heatmap(
      na_col = "white",
    # km = 3,
    cluster_columns = F,
    #cluster_rows = F,
    show_row_dend = F,
    show_column_dend = F,
    show_row_names = T,
    row_names_gp = grid::gpar(fontsize = 8.5),  # Adjust font size here
    column_names_gp = grid::gpar(fontsize = 11),
    rect_gp = grid::gpar(col = "lightgrey", lwd = 0.1), ##3A3A3A
    col = circlize::colorRamp2(c(-2, 0, 2), c("dodgerblue4", "white", "darkred")),
    name = "log2FC",
    column_order = column_order#,
    # row_order = TF_order
  )

p_tf

```
